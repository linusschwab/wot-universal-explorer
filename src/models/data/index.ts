export {Data} from './Data'
export {InputData} from './InputData'
export {OutputData} from './OutputData'
