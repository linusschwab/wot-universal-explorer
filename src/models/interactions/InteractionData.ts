export class InteractionData {

    public timestamp: number;
    public data: any;

    constructor(data: any) {
        this.timestamp = Date.now();
        this.data = data;
    }

    public toString() {
        return {
            time: this.date,
            data: this.data
        }
    }

    public equals(data: InteractionData) {
        return JSON.stringify(this.data) === JSON.stringify(data.data);
    }

    get time() {
        return new Date(this.timestamp).toLocaleTimeString('de-CH');
    }

    get date() {
        return new Date(this.timestamp).toLocaleString('de-CH');
    }
}
