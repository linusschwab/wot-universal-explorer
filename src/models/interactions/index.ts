export {InteractionPattern} from './InteractionPattern';
export {Action} from './Action';
export {Event} from './Event';
export {Property} from './Property';
