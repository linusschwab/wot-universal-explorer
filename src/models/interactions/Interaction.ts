import {Link} from "../links";
import {OutputData} from "../data";

export class Interaction {

    public name: string;
    public base: string;
    public links: Link[];
    public outputData: OutputData;

    constructor(name: string) {
        this.name = name;
        this.links = [];
    }

    public registerLink(link: Link) {
        this.links.push(link);
        link.interaction = this;
    }

    public toString() {
        return this.name + ' interaction';
    }

}
