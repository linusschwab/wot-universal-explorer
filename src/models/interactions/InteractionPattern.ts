import * as WebSocket from "ws";
import * as getSlug from "speakingurl";

import {Link, Operation} from "../links";
import {Thing} from "../thing";
import {InteractionData} from "./InteractionData";
import {InteractionError} from "../../tools/errors";


export abstract class InteractionPattern {

    public name: string;
    public description: string;

    public thing: Thing;
    public base: string;

    public links: Link[];
    public url: string;
    public operations: Operation[];

    public data: InteractionData[];
    public subscribers: WebSocket[];

    constructor(name: string) {
        this.name = name;
        this.links = [];
        this.data = [];
        this.subscribers = [];
    }

    public registerLink(link: Link) {
        this.links.push(link);
        link.interaction = this;
    }

    public subscribe(ws: WebSocket) {
        if (this.subscribers.includes(ws)) {
            throw new InteractionError('Already subscribed to interaction');
        }

        this.subscribers.push(ws);
    }

    public unsubscribe(ws: WebSocket) {
        if (!this.subscribers.includes(ws)) {
            throw new InteractionError('Not subscribed to interaction');
        }

        let index = this.subscribers.indexOf(ws);
        this.subscribers.splice(index, 1);
    }

    public abstract async notifySubscribers(data: InteractionData): Promise<void>;

    public async storeData(data: InteractionData) {
        // Delete old items if too many
        if (this.data.length > 1000) {
            this.data.splice(0, 1);
        }

        this.data.push(data);
    }

    public newData(data: InteractionData): boolean {
        return this.data.length == 0 || !data.equals(this.data[this.data.length-1])
    }

    public abstract toString(): string;

    get slug() {
        return getSlug(this.name);
    }
}
