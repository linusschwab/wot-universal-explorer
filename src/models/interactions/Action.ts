import {InteractionPattern} from "./InteractionPattern";
import {InputSchema, OutputSchema} from "../schema";
import {Operation} from "../links";


export class Action extends InteractionPattern {

    public outputSchema: OutputSchema;
    public inputSchema: InputSchema;

    constructor(name: string, outputSchema: OutputSchema, inputSchema: InputSchema) {
        super(name);
        this.outputSchema = outputSchema;
        this.inputSchema = inputSchema;
    }

    public async invoke(data: any = null) {
        // TODO: Choose correct link
        return this.links[0].execute(data);
    }

    public toString() {
        return 'Action ' + this.name;
    }

    public get url() {
        return '/' + this.thing.slug + '/actions/' + this.slug;
    }

    public get operations(): Operation[] {
        return [new Operation('post', 'Invoke ' + this.toString(), this)];
    }
}
