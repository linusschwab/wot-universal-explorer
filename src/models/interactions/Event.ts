import {InteractionPattern} from "./InteractionPattern";
import {DataSchema} from "../schema";
import {Operation} from "../links";
import {InteractionData} from "./InteractionData";
import {TimeoutError} from "../../tools/errors";


export class Event extends InteractionPattern {

    public schema: DataSchema;

    constructor(name: string, schema: DataSchema) {
        super(name);
        this.schema = schema;
    }

    public async poll() {
        // TODO: Choose correct link
        return this.links[0].execute(null);
    }

    public async update() {
        // Only poll if regular event link
        if (this.links.length === 0) {
            return;
        }

        try {
            const data = new InteractionData(await this.poll());

            // Check if data changed
            if (this.newData(data)) {
                this.storeData(data);
                this.notifySubscribers(data);
            }
        } catch (e) {
            if (e instanceof TimeoutError === false) {
                throw e;
            }
        }
    }

    public getData(newerThan: number = 0, limit: number = 0) {
        let newData: InteractionData[] = [];
        if (newerThan > 0) {
            for (let data of this.data) {
                if (data.timestamp > newerThan) {
                    newData.push(data);
                }
            }
        } else {
            newData = this.data;
        }
        return newData.slice(-limit);
    }

    public async notifySubscribers(data: InteractionData) {
        for (let ws of this.subscribers) {
            ws.send(JSON.stringify({
                messageType: 'event',
                data: data.toString()
            }));
        }
    };

    public toString() {
        return 'Event ' + this.name;
    }

    public get url() {
        return '/' + this.thing.slug + '/events/' + this.slug;
    }

    public get operations(): Operation[] {
        return [new Operation('get', 'List of ' + this.name + ' Event data', this)];
    }
}
