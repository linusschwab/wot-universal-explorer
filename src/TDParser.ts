import {Thing} from "./models/thing/Thing";
import {Action, Event, Interaction, Property} from "./models/interactions";
import {InputData, OutputData} from "./models/data";
import {HTTPLink} from "./models/links";


export class TDParser {

    public static parse(td: string) {
        const obj = JSON.parse(td);

        let thing = new Thing(obj.name, obj.base);

        for (let iobj of obj.interactions) {
            let interaction = this.parseInteraction(iobj);

            for (let lobj of iobj.links) {
                let link = this.parseLink(lobj, thing.base);
                interaction.registerLink(link);
            }

            thing.registerInteraction(interaction);
        }

        return thing;
    }

    // Interactions
    private static parseInteraction(iobj: any) {
        let type = iobj['@type'][0];

        switch(type) {
            case 'Action':
                return this.parseAction(iobj);
            case 'Event':
                return this.parseEvent(iobj);
            case 'Property':
                return this.parseProperty(iobj);
            default:
                let name = iobj.name;
                return new Interaction(name);
        }
    }

    private static parseAction(iobj: any) {
        let name = iobj.name;

        let outputData = null;
        if ('outputData' in iobj) {
            outputData = new OutputData('data', iobj.outputData.valueType.type, false, false);
        }

        let inputData = null;
        if ('inputData' in iobj) {
            inputData = new InputData('data', iobj.inputData.valueType.type, true);
        }

        return new Action(name, outputData, inputData);
    }

    private static parseEvent(iobj: any) {
        let name = iobj.name;
        let outputData = new InputData('data', iobj.outputData.valueType.type, false);

        return new Event(name, outputData);
    }

    private static parseProperty(iobj: any) {
        let name = iobj.name;

        let writable = false;
        if ('writable' in iobj) {
            writable = iobj.writable;
        }

        let outputData = new OutputData('data', iobj.outputData.valueType.type, writable, false);

        return new Property(name, outputData, writable);
    }

    // Links
    private static parseLink(lobj: any, base: string) {
        let href = lobj.href;
        let mediaType = lobj.mediaType;
        // TODO: Support other link types
        return new HTTPLink(href, base, mediaType);
    }
}
