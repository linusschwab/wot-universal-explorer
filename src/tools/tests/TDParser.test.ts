import * as fs from "fs";
import {TDParser} from "../TDParser";
import {Action} from "../../models/interactions";


test('td without base and 1 action with multiple links', () => {
    const tdLamp = fs.readFileSync(__dirname + '/__data__/td-sample-lamp.jsonld', 'utf8');
    const thing = TDParser.parse(tdLamp);

    // Basic data
    expect(thing.name).toBe('myMasterLED');
    expect(thing.type).toBe('Lamp');
    expect(thing.base).toBe('');
    expect(thing.interactions).toHaveLength(1);

    // Interactions
    const action = thing.interactions[0];
    expect(action).toBeInstanceOf(Action);
    expect(action.name).toBe('myMasterOnOff');
    expect(action.links).toHaveLength(2);
    expect(action.links[0].toString()).toBe('coap://www.example.com:5683/master');
    expect(action.links[1].toString()).toBe('http://www.example.com:80/master');

    // Data schema
    const actionInputSchema = (<Action>action).inputSchema;
    expect(actionInputSchema.type).toBe('boolean');
    expect(actionInputSchema.writable).toBe(true);

    const actionOutputSchema = (<Action>action).outputSchema;
    expect(actionOutputSchema).toBeNull();
});

test('td with base, 2 properties and 2 events', () => {
    const tdSensor = fs.readFileSync(__dirname + '/__data__/td-sample-sensor.jsonld', 'utf8');
    const thing = TDParser.parse(tdSensor);

    // Basic data
    expect(thing.name).toBe('myTempSensor');
    expect(thing.type).toBe('Sensor');
    expect(thing.base).toBe('coap:///www.example.com:5683/temp/');
    expect(thing.interactions).toHaveLength(4);

    // Interactions
    expect(thing.interactions).toMatchSnapshot()
});
