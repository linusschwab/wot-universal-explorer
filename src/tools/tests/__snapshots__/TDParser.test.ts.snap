// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`td with base, 2 properties and 2 events 1`] = `
Array [
  Property {
    "base": "coap:///www.example.com:5683/temp/",
    "links": Array [
      HTTPLink {
        "host": "coap:///www.example.com:5683/temp/",
        "href": "val",
        "http": [Function],
        "interaction": [Circular],
        "mediaType": "application/json",
      },
    ],
    "name": "myTemp",
    "observable": true,
    "schema": DataSchema {
      "required": false,
      "type": "number",
      "writable": false,
    },
    "thing": Thing {
      "base": "coap:///www.example.com:5683/temp/",
      "interactions": [Circular],
      "name": "myTempSensor",
      "type": "Sensor",
    },
    "writable": false,
  },
  Property {
    "base": "coap:///www.example.com:5683/temp/",
    "links": Array [
      HTTPLink {
        "host": "coap:///www.example.com:5683/temp/",
        "href": "thr",
        "http": [Function],
        "interaction": [Circular],
        "mediaType": "application/json",
      },
    ],
    "name": "myThreshold",
    "observable": false,
    "schema": DataSchema {
      "required": false,
      "type": "number",
      "writable": true,
    },
    "thing": Thing {
      "base": "coap:///www.example.com:5683/temp/",
      "interactions": [Circular],
      "name": "myTempSensor",
      "type": "Sensor",
    },
    "writable": true,
  },
  Event {
    "base": "coap:///www.example.com:5683/temp/",
    "links": Array [
      HTTPLink {
        "host": "coap:///www.example.com:5683/temp/",
        "href": "val/changed",
        "http": [Function],
        "interaction": [Circular],
        "mediaType": "application/json",
      },
    ],
    "name": "myChange",
    "schema": DataSchema {
      "required": false,
      "type": "number",
      "writable": false,
    },
    "thing": Thing {
      "base": "coap:///www.example.com:5683/temp/",
      "interactions": [Circular],
      "name": "myTempSensor",
      "type": "Sensor",
    },
  },
  Event {
    "base": "coap:///www.example.com:5683/temp/",
    "links": Array [
      HTTPLink {
        "host": "coap:///www.example.com:5683/temp/",
        "href": "val/high",
        "http": [Function],
        "interaction": [Circular],
        "mediaType": "application/json",
      },
    ],
    "name": "myWarning",
    "schema": DataSchema {
      "required": false,
      "type": "number",
      "writable": false,
    },
    "thing": Thing {
      "base": "coap:///www.example.com:5683/temp/",
      "interactions": [Circular],
      "name": "myTempSensor",
      "type": "Sensor",
    },
  },
]
`;
